llarp/net/route.cpp

```diff
@@ -338,28 +340,28 @@ namespace llarp::net
   {
     std::vector<std::string> gateways;
 #ifdef __linux__
-
-    FILE* p = popen("ip route", "r");
-    if (p == nullptr)
-      return gateways;
-    char* line = nullptr;
-    size_t len = 0;
-    ssize_t read = 0;
-    while ((read = getline(&line, &len, p)) != -1)
-    {
-      std::string line_str(line, len);
-      std::vector<std::string> words;
-      std::istringstream instr(line_str);
-      for (std::string word; std::getline(instr, word, ' ');)
+    // return true when it's all zeros
+    const auto is_zero = [](const auto& str) {
+      for (const auto& ch : str)
       {
-        words.emplace_back(std::move(word));
+        if (ch != '0')
+          return false;
       }
-      if (words[0] == "default" and words[1] == "via" and words[3] == "dev" and words[4] != ifname)
+      return true;
+    };
+    std::ifstream inf("/proc/net/route");
+    for (std::string line; std::getline(inf, line);)
+    {
+      const auto parts = split(std::string_view{line}, '\t');
+      if (is_zero(parts[1]) and parts[0] != ifname)
       {
-        gateways.emplace_back(std::move(words[2]));
+        nuint32_t x{};
+        const auto& ip = parts[2];
+        std::from_chars(ip.begin(), ip.end(), x.n, 16);
+        gateways.emplace_back(x.ToString());
```

Comment:
This is weird and it was not at all obvious to me why it works.

The hex value (e.g. `0100000A` for 10.0.0.1) is the little endian byte representation (*not* a hex-encoded 32-byte value) but `from_chars` always interprets values as a hex-encoded values (which are implicitly big-endian), so it ends up doing the byte swap -- but doing a byte swap implicitly inside a text-to-string converter is *really* unintuitive.  I think this would be clearer:
```C++
if (ip.size() == 2*sizeof(uint32_t) && lokimq::is_hex(ip))
{
  huint32_t x;
  lokimq::from_hex(ip.begin(), ip.end(), reinterpret_cast<char*>(&x.h));
  gateways.emplace_back(x.ToString());
}
```

Reply:
done